src/crypto/chacha20.cpp

```diff
@@ -71,7 +71,7 @@ void ChaCha20::Seek(uint64_t pos)
     input[13] = pos >> 32;
 }
 
-void ChaCha20::Output(unsigned char* c, size_t bytes)
+void ChaCha20::Output(const unsigned char* m, unsigned char* c, size_t bytes)
```

Comment:
I think it would be better to duplicate the logic here and turn the stream cipher version into a separate function.

As is, this will do double work in outputting both, and introduce an unnecessary branch in the output version.

Reply:
Good point. Adapted the duplication approach.

Reply:
What's the reasoning here? Performance?

How does the version in the refrence implementation compare?

```c
void ECRYPT_keystream_bytes(ECRYPT_ctx *x,u8 *stream,u32 bytes)
{
  u32 i;
  for (i = 0;i < bytes;++i) stream[i] = 0;
  ECRYPT_encrypt_bytes(x,stream,stream,bytes);
}
```

ie setting `Output(c, bytes)` to call `Crypt(c, c, bytes)`?

Reply:
I had it in the initial version like @jnewbery just proposed.
After @sipa recommended to separate it, I thought that a strict separation may be beneficial for verification and later optimizations. But maybe @sipa can clarify...